Refactor code for Extension API doc (#407)

FabienLelaquais opened the pull request.

Commit:
Few items refactored

FredLL-Avaiga committed:
Add get_js_module_name

FredLL-Avaiga committed:
Supporting several libraries with the same "backend" name

FredLL-Avaiga committed:
renaming user_element.py to library.py

FredLL-Avaiga committed:
Make xml element accessible to be able to define a child node as text

Commit:
Some small refactoring items.

Commit:
Merge branch 'doc/#407-extension-api' of github.com:Avaiga/taipy-gui into doc/#407-extension-api

FredLL-Avaiga commented on src/taipy/gui/extension/library.py:
```diff
@@ -22,74 +22,78 @@
+    The definition of a custom visual element.
 
+    The definition of an element is made of its names, its properties, and
+    the
```
Unfinished business?

FabienLelaquais replied:
Indeed. I was waiting for your own changes before adding anything can could turn wrong.

FredLL-Avaiga commented on src/taipy/gui/extension/library.py:
```diff
@@ -162,52 +166,71 @@ def render(self, gui: "Gui", properties: t.Dict[str, t.Any], hash_names: t.Dict[
     """
-    TODO
-    This class needs to be inherited to define a new library.
+    An library of user-defined visual elements.
```
A

FabienLelaquais replied:
✔️

FredLL-Avaiga commented on src/taipy/gui/extension/library.py:
```diff
@@ -162,52 +166,71 @@ def render(self, gui: "Gui", properties: t.Dict[str, t.Any], hash_names: t.Dict[
+        Must be unique on the browser window scope.
+
+        """
+        return self.get_name()
```
Use Carmel case?

FabienLelaquais replied:
Indeed.

FredLL-Avaiga reviewed (approved):
A few comments

Commit:
Comments integrated.

FabienLelaquais committed:
Merge branch 'develop' into doc/#407-extension-api

Commit:
Fix tests

Commit:
Merge branch 'doc/#407-extension-api' of github.com:Avaiga/taipy-gui into doc/#407-extension-api

Commit:
Fix the fix...

FredLL-Avaiga committed:
add ssl_context

FredLL-Avaiga committed:
- do not publish Builder yet
- give a simple way to render xHtml

dinhlongviolin1 committed:
Merge branch 'develop' into doc/#407-extension-api

dinhlongviolin1 committed:
Merge branch 'develop' into doc/#407-extension-api

Commit:
Extension factory refactor.

FredLL-Avaiga commented on src/taipy/gui/extension/library.py:
```diff
@@ -34,94 +34,93 @@ class ElementProperty:
-    The definition of an element is made of its names, its properties, and
-    the
-    TODO
+    An element is defined by its properties (name, type and default value) and
```
name ?

FabienLelaquais replied:
Correct. No name!

FabienLelaquais replied:
Yes name! The name... of the properties.

FredLL-Avaiga commented on src/taipy/gui/gui.py:
```diff
@@ -271,8 +271,12 @@ def __init__(
+        Arguments:
+            library: The custom visual element library to add to this library.
+
+        TODO: What if we add two libraries with the same name?
```
done

FredLL-Avaiga reviewed (approved):
let's do this

Commit:
Doc

FredLL-Avaiga committed:
#424 remove register function and make react components unique

FredLL-Avaiga committed:
fix tests

FredLL-Avaiga committed:
define a default implementation for library.get_resource

FredLL-Avaiga committed:
PathLib.is_relative_to is new to Python 3.9

Commit:
- Rename taipy-vendor* to taipy-gui-dependencies
- JS-less extensions are first class citizens
- Added ElementLibrary.get_resource_url()

FredLL-Avaiga committed:
#425 multiple libraries with same name
